src/HaskellWorks/Data/Network/Ip/Block.hs

```diff
@@ -0,0 +1,68 @@
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DuplicateRecordFields      #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE InstanceSigs               #-}
+{-# LANGUAGE OverloadedStrings          #-}
+
+module HaskellWorks.Data.Network.Ip.Block
+  ( IpBlock(..)
+  , isValidIpBlock
+  , firstAddress
+  , lastAddress
+  ) where
+
+import Control.Applicative
+import Control.Monad
+import Data.Char
+import Data.Maybe
+import Data.Word
+import GHC.Generics
+import HaskellWorks.Data.Bits.BitWise
+import Text.Read
+
+import qualified Data.Attoparsec.Text                  as AP
+import qualified Data.Bits                             as B
+import qualified Data.Text                             as T
+import qualified HaskellWorks.Data.Network.Ip          as Z
+import qualified HaskellWorks.Data.Network.Ip.Internal as I
+import qualified HaskellWorks.Data.Network.Ip.Ipv4     as I4
+import qualified HaskellWorks.Data.Network.Ip.Ipv6     as I6
+import qualified Text.ParserCombinators.ReadPrec       as RP
+
+newtype IpBlock = IpBlock
+  { block :: Either I4.Ipv4Block I6.Ipv6Block
+  } deriving (Eq, Ord, Generic)
+
+
+instance Show IpBlock where
+  showsPrec _ (IpBlock (Left ipv4Block))  = shows ipv4Block
+  showsPrec _ (IpBlock (Right ipv6Block)) = shows ipv6Block
+
+instance Read IpBlock where
+  readsPrec _ s =
+    case readMaybe s :: Maybe I4.Ipv4Block of
+      Just ipv4 -> [(IpBlock (Left ipv4), "")]
+
+      Nothing ->
+        case readMaybe s :: Maybe I6.Ipv6Block of
+          Just ipv6 -> [(IpBlock (Right ipv6), "")]
+          Nothing   -> []
+
+isValidIpBlock :: IpBlock -> Bool
+isValidIpBlock (IpBlock (Left _))    = True -- TODO
```

Comment:
Why is this `TODO`?

Reply:
I don't know, it is like that in the original source project as well.